ddtrace/tracer/writer.go

```diff
@@ -0,0 +1,302 @@
+		prioritySampling: s,
+	}
+}
+
+func (h *agentTraceWriter) add(trace []*span) {
+	if err := h.payload.push(trace); err != nil {
+		h.config.statsd.Incr("datadog.tracer.traces_dropped", []string{"reason:encoding_error"}, 1)
+		log.Error("error encoding msgpack: %v", err)
+	}
+	if h.payload.size() > payloadSizeLimit {
+		h.config.statsd.Incr("datadog.tracer.flush_triggered", []string{"reason:size"}, 1)
+		h.flush()
+	}
+}
+
+func (h *agentTraceWriter) stop() {
+	h.wg.Wait()
+}
+
+// flush will push any currently buffered traces to the server.
+func (h *agentTraceWriter) flush() {
+	if h.payload.itemCount() == 0 {
+		return
+	}
+	h.wg.Add(1)
+	h.climit <- struct{}{}
+	go func(p *payload) {
+		defer func(start time.Time) {
+			<-h.climit
+			h.wg.Done()
+			h.config.statsd.Timing("datadog.tracer.flush_duration", time.Since(start), nil, 1)
+		}(time.Now())
+		size, count := p.size(), p.itemCount()
+		log.Debug("Sending payload: size: %d traces: %d\n", size, count)
+		rc, err := h.config.transport.send(p)
+		if err != nil {
+			h.config.statsd.Count("datadog.tracer.traces_dropped", int64(count), []string{"reason:send_failed"}, 1)
+			log.Error("lost %d traces: %v", count, err)
+		} else {
+			h.config.statsd.Count("datadog.tracer.flush_bytes", int64(size), nil, 1)
+			h.config.statsd.Count("datadog.tracer.flush_traces", int64(count), nil, 1)
+			if err := h.prioritySampling.readRatesJSON(rc); err != nil {
+				h.config.statsd.Incr("datadog.tracer.decode_error", nil, 1)
+			}
+		}
+	}(h.payload)
+	h.payload = newPayload()
+}
+
+// logTraceWriter encodes traces into a format understood by the Datadog Forwarder
+// (https://github.com/DataDog/datadog-serverless-functions/tree/master/aws/logs_monitoring)
+// and writes them to os.Stdout. This is used to send traces from an AWS Lambda environment.
+type logTraceWriter struct {
+	config    *config
+	buf       bytes.Buffer
+	hasTraces bool
+	w         io.Writer
+}
+
+func newLogTraceWriter(c *config) *logTraceWriter {
+	w := &logTraceWriter{
+		config: c,
+		w:      os.Stdout,
+	}
+	w.resetBuffer()
+	return w
+}
+
+const (
+	// maxFloatLength is the maximum length that a string encoded by encodeFloat will be.
+	maxFloatLength = 24
+
+	// logBufferSuffix is the final string that the trace writer has to append to a buffer to close
+	// the JSON.
+	logBufferSuffix = "]}\n"
+
+	// logBufferLimit is the maximum size log line allowed by cloudwatch
+	logBufferLimit = 256 * 1024
+)
+
+func (h *logTraceWriter) resetBuffer() {
+	h.buf.Reset()
+	h.buf.WriteString(`{"traces": [`)
+	h.hasTraces = false
+}
+
+// encodeFloat correctly encodes float64 into the JSON format followed by ES6.
+// This code is reworked from Go's encoding/json package
+// (https://github.com/golang/go/blob/go1.15/src/encoding/json/encode.go#L573)
+//
+// One important departure from encoding/json is that infinities are encoded as
+// math.MaxFloat64 and -math.MaxFloat64, and NaNs are encoded as NaN.
```

Comment:
I need your opinion on this, @gbbr.
I'm not confident this is the best way to go. NaN`s are filtered out below in `encodeSpan` since they are not valid JSON values. I'm approximating Infinities to `MaxFloat64` and `-MaxFloat64`.

Msgpack supports NaN and +-Infinity natively, and they show up as tags with null values in the UI. What do you think?

Reply:
> Msgpack supports NaN and +-Infinity natively, and they show up as tags with null values in the UI.

I think that's your answer :)

Reply:
For now I'm setting them to `null`. Unfortunately it looks like that is decoded as `0` by go's json decoder.
Still waiting for definitive answers from the downstream projects.